pkg/agent/cluster/controller.go

```diff
@@ -0,0 +1,157 @@
+	scheme                     *runtime.Scheme
+	volumeAttachmentController crd.VolumeAttachmentController
+	flexvolumeController       *flexvolume.FlexvolumeController
+}
+
+func NewClusterController(context *clusterd.Context, volumeAttachmentClient rest.Interface,
+	manager flexvolume.VolumeManager) (*ClusterController, error) {
+
+	volumeAttachmentController, err := crd.NewVolumeAttachmentController(context.Clientset, volumeAttachmentClient)
+	if err != nil {
+		return nil, err
+	}
+
+	flexvolumeController, err := flexvolume.NewFlexvolumeController(context, volumeAttachmentClient, manager)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ClusterController{
+		context:                    context,
+		volumeAttachmentController: volumeAttachmentController,
+		flexvolumeController:       flexvolumeController,
+	}, nil
+}
+
+func (c *ClusterController) StartWatch(namespace string, stopCh chan struct{}) error {
+
+	customResourceClient, scheme, err := opkit.NewHTTPClient(k8sutil.CustomResourceGroup, k8sutil.V1Alpha1, opcluster.SchemeBuilder)
+	if err != nil {
+		return fmt.Errorf("failed to get a k8s client for watching cluster resources: %v", err)
+	}
+	c.scheme = scheme
+
+	resourceHandlerFuncs := cache.ResourceEventHandlerFuncs{
+		DeleteFunc: c.onDelete,
+	}
+
+	watcher := opkit.NewWatcher(opcluster.ClusterResource, namespace, resourceHandlerFuncs, customResourceClient)
+	go watcher.Watch(&opcluster.Cluster{}, stopCh)
+	return nil
+}
+
+func (c *ClusterController) onDelete(obj interface{}) {
+	clusterOrig := obj.(*opcluster.Cluster)
+
+	// NEVER modify objects from the store. It's a read-only, local cache.
+	// Use scheme.Copy() to make a deep copy of original object.
+	copyObj, err := c.scheme.Copy(clusterOrig)
+	if err != nil {
+		logger.Errorf("failed to create a deep copy of cluster object: %v\n", err)
+		return
+	}
+	cluster := copyObj.(*opcluster.Cluster)
+
+	node := os.Getenv(k8sutil.NodeNameEnvVar)
+	agentNamespace := os.Getenv(k8sutil.PodNamespaceEnvVar)
+	logger.Infof("cluster in namespace %s is being deleted, agent on node %s will attempt to clean up.", cluster.Namespace, node)
+
+	// TODO: filter this List operation by node name and cluster namespace on the server side
+	vols, err := c.volumeAttachmentController.List(agentNamespace)
+	if err != nil {
+		logger.Errorf("failed to get volume attachments for agent namespace %s: %+v", agentNamespace, err)
+	}
+
+	// find volume attachments in the deleted cluster that are attached to this node
+	for _, vol := range vols.Items {
+		for _, a := range vol.Attachments {
+			if a.Node == node && a.ClusterName == cluster.Namespace {
+				logger.Infof("volume %s has an attachment belonging to deleted cluster %s, will clean it up now. mountDir: %s",
+					vol.Name, cluster.Namespace, a.MountDir)
+
+				// start a goroutine to perform the cleanup of this volume attachment asynchronously.
+				// if one cleanup hangs, it will not affect the others.
+				go func() {
```

Comment:
This is pretty smart to do the cleanup in parallel. Hopefully the cleanup is performed before the graceful period.